tests/e2e/tests/pilot/mirror_test.go

```diff
@@ -0,0 +1,150 @@
+			percentage: 50.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-10%",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-10.yaml",
+			percentage: 10.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-80%",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-80.yaml",
+			percentage: 80.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-0%",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-0.yaml",
+			percentage: 0.0,
+			threshold:  0.0,
+		},
+	}
+
+	cfg := &deployableConfig{
+		Namespace:  tc.Kube.Namespace,
+		kubeconfig: tc.Kube.KubeConfig,
+	}
+	if err := cfg.Setup(); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_ = cfg.Teardown()
+	}()
+
+	podsC := getAppPods(t, "c")
+	podsB := getAppPods(t, "b")
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+
+			if err := util.KubeApply(tc.Kube.Namespace, c.yamlFile, tc.Kube.KubeConfig); err != nil {
+				t.Fatalf("Error applying mirror yaml file: %v", err)
+			}
+			time.Sleep(5 * time.Second)
+			defer func() {
+				if err := util.KubeDelete(tc.Kube.Namespace, c.yamlFile, tc.Kube.KubeConfig); err != nil {
+					t.Fatalf("Error deleting mirror yaml file: %v", err)
+				}
+			}()
+
+			reqURL := "http://c/a"
+			testID := randomString(16)
+			count := 0
+
+			for cluster := range tc.Kube.Clusters {
+				for i := 1; i <= 100; i++ {
+					resp := ClientRequest(cluster, "a", reqURL, 1, fmt.Sprintf("--key X-Request-Id --val %d-%s", i, testID))
+					if len(resp.ID) > 0 {
+						count++
+					}
+				}
+			}
+
+			runRetriableTest(t, "verifyLogs", defaultRetryBudget, func() error {
+				countC := logCount(podsC, testID)
+				countB := logCount(podsB, testID)
+				ratio := (countB / countC) * 100
+
+				if ratio < (c.percentage-c.threshold) || ratio > (c.percentage+c.threshold) {
+					log.Errorf("mirror ratio is not desired. Desired=%v (threshold=%v) testId=%v countLocal=%v countC=%v, countB=%v, ratio=%v",
+						c.percentage, c.threshold, testID, count, countC, countB, ratio)
+					return errAgain
+				}
+
+				log.Infof("mirror ratio is OK. Desired=%v (threshold=%v) testId=%v countLocal=%v countC=%v, countB=%v, ratio=%v",
+					c.percentage, c.threshold, testID, count, countC, countB, ratio)
+				return nil
+			})
+
+		})
+	}
+}
+
+func logCount(pods map[string][]string, testID string) float64 {
+	var logs string
+	for c, cPods := range pods {
+		for _, pod := range cPods {
+			logs += util.GetPodLogs(tc.Kube.Namespace, pod, inject.ProxyContainerName, false, false, tc.Kube.Clusters[c])
+		}
+	}
+	return float64(strings.Count(logs, testID))
+}
+
+const letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+// RandomString returns a random string of size n
+func randomString(n int) string {
+	rand.Seed(time.Now().UnixNano())
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = letters[rand.Int63()%int64(len(letters))]
+	}
+	return string(b)
+}
```

Comment:
Should this go under the `test/utils` package?

Reply:
done